### DISCUSSION TITLE
Using Oboe with Kotlin

### DISCUSSION THREAD

**Discussion by Ateros4me** [google/oboe#1418] [Q&A] [answered]
How do I get the Oboe library to be loaded by Kotlin in Android Studio?
I’ve implemented the recommendations in https://www.youtube.com/watch?v=csfHAbr5ilI but when I do an assignment to an object like the following the oboe is highlighted in red and typing alt-cr doesn’t import the library

val audioTrack = oboe.AudioStream( etc…

I’m probably being a bit dumb but any help would be useful.

I have attached the CMakeLists.txt file
[CMakeLists.txt](https://github.com/google/oboe/files/7357895/CMakeLists.txt)

I can see the Gradle builder loading in the oboe 1.6.1 in android studio but i can’t access the oboe Audio Stream via Kotlin

**Comment by android-usb** [comment]
Have a look at the DrumThumper demo. It's UI is written in Kotlin.

**Reply by Ateros4me** [reply]
Thanks for the reply, have you a link to drumthumper as i can’t seem to find it?

**Reply by Ateros4me** [reply]
maybe it would be better if i say what i’m trying to do.
I just want to receive audio in, pass it through an equaliser then send the audio out to headphones, however i want the latency to be sufficiently low that the human ear doesn’t detect it.
Are there any examples of something like that?
Thanks

**Reply by dturner** [reply]
For this kind of use case you're best off using the `LiveEffect` which "echos" (copies) the input to the output (code here: https://github.com/google/oboe/blob/main/samples/LiveEffect/src/main/cpp/FullDuplexPass.h#L46). You can add your equalizer code before copying the input to the output. This should all be done in C++.

**Comment by android-usb** [comment]
DrumThumper is in the "samples" subproject in the Oboe repository.

https://github.com/google/oboe/tree/main/samples/drumthumper

- Paul

On Tue, Oct 19, 2021 at 2:00 AM Ateros4me ***@***.***> wrote:

> Thanks for the reply, have you a link to drmthumper as i can’t seem to
> find it?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/oboe/discussions/1418#discussioncomment-1500714>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24N6JDQL6ALLKY6X6UP2LUHUQQNANCNFSM5GDZ5OAA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Paul McLean
Software Engineer
Android Audio Engineering
303 579 0404

**Reply by Ateros4me** [reply]
Hi Paul
Is there any reason why this sample code can't be cloned with git as the "code" button isn't visible?
Thanks
Chris

**Reply by Ateros4me** [reply]
Sorry, me being a bit dumb, found it in the clone of the oboe-main :)

**Reply by Ateros4me** [reply]
Sorry, but this seems to me to be a massive botch.
So Kotlin ends up calling Java which then calls C++ rather than fixing the issue and getting Kotlin to call an Oboe routine directly rather like the current way AudioTrack and AudioRecord etc are called from Kotlin.
Is there any plan to fix this for Kotlin properly?
Regards
Chris

**Reply by dturner** [reply]
Hi Chris, 

I'm not sure what you mean by "fixing the issue". Kotlin compiles to Java bytecode so can call the Oboe methods in the same way that Java does, via the Java Native Interface (JNI). Aside from some syntax changes, calling C++ methods from Kotlin is exactly the same as calling C++ methods from Java.  

`AudioTrack` and `AudioRecord` are part of the Java API so they are available from Kotlin without going through JNI. 

Oboe is a C++ library and operates on the lower-level C libraries in Android (like OpenSL ES and AAudio). It does this for performance reasons. We recommend that you keep your UI in Java/Kotlin and do your audio processing in C++. 

Hope that helps, 

Don

**Reply by Ateros4me** [reply]
Hi Don

By “fixing” i meant, could it all be hidden under the hood like AudioTrack in Kotlin. I do appreciate everyones help, I’m just finding it daunting, as i felt i’d made progress only to find there is far more work than i originally thought.

Regards
Chris

**Comment by android-usb** [comment, marked as answer]
The whole point of Oboe is to provide a high-performance **C++** audio API, particularly getting all of the audio processing the C/C++ layer. Wrapping in it a Java/Kotlin layer would eliminate this benefit (i.e. it would become a Java/Kotlin API).

If the Java AudioTrack API is sufficient for your purposes then you should use that. Since I perceive you want high performance (for real-time audio processing), you want to be working in the C/C++ layer.

The recommended way to use Oboe in an application is to write just the UI layer in Java/Kotlin and have all of the audio processing in C/C++. It is certainly the case that you need to know how to write JNI (Java Native Interface) code, but that is sort-of the price of entry.

I'm not as familiar with the FXLab sample, but DrumThumper does show how to use Oboe from a Kotlin app and handle all of the audio processing in the high-performance C/C++ layer.

I hope that helps.
Regards,
Paul

**Comment by Ateros4me** [comment]
Hi Don,

i’ve tried using the DuplexEngine from the fxlab in the oboe clone checkout, but i keep getting the following error
no template named ‘variant’ in namespace ‘std’ :42
in DuplexEngine.h
I have the following set in my app build.gradle
minSdkVersion 21
targetSdkVersion 25
compileSdkVersion 31

I can find the variant file under Android/Sdk/ndk/21.4.7075529/sources/cxx-stl/llvm-libc++/include
I also added c++17 in to the build.gradle as follows
externalNativeBuild{
    cmake {
               arguments “-DANDROID_STL=c++_shared”, “-std=c++17”
                 }
 }
 
but this didn’t fix the problem
Do you have any suggestions?

Chris

**Reply by dturner** [reply]
You should have posted this as a new question but I'll answer....

You'll need to specify C++17 in your `CMakeLists.txt` as per the fxlab example here: https://github.com/google/oboe/blob/4fa162bc5f1949f4130e1ee43ae237e8359b8d9c/apps/fxlab/app/CMakeLists.txt#L9

If you have follow up comments/questions please create a new question and copy/paste the contents of this thread into it.

**Reply by Ateros4me** [reply]
Sorry, i thought it was continuing on from the previous discussion. Thanks for replying
